File: sdks/go/tasks.go
package decktools

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

const (
	sseRetryInterval = 5 * time.Second
	sseMaxRetries    = 100
)

type TasksClient struct {
	http  *httpClient
	files *FilesClient
}

func (t *TasksClient) Create(ctx context.Context, params CreateTaskParams) (*Task, error) {
	spaceID, err := t.http.ResolveSpaceID(ctx, params.SpaceID)
	if err != nil {
		return nil, err
	}
	prepared, err := t.prepareTaskFiles(ctx, params)
	if err != nil {
		return nil, err
	}

	var totalBytes int64
	for _, file := range prepared {
		totalBytes += file.Bytes
	}
	canInline := len(prepared) > 0 && len(params.FileIDs) == 0 && totalBytes < InlineTaskFilesMaxBytes
	if canInline {
		return t.createWithInlineFiles(ctx, spaceID, params, prepared)
	}

	fileIDs, err := t.resolveFileIDs(ctx, spaceID, params, prepared)
	if err != nil {
		return nil, err
	}
	payload := map[string]any{
		"fileIds": fileIDs,
		"type":    params.Type,
		"params":  params.Params,
	}
	if payload["params"] == nil {
		payload["params"] = map[string]any{}
	}
	if spaceID != "" {
		payload["spaceId"] = spaceID
	}
	if params.Name != "" {
		payload["name"] = params.Name
	}

	var task Task
	if _, err := t.http.postJSON(ctx, "/tools/tasks", payload, &task); err != nil {
		return nil, err
	}
	return &task, nil
}

func (t *TasksClient) createWithInlineFiles(ctx context.Context, spaceID string, params CreateTaskParams, files []*normalizedUpload) (*Task, error) {
	var body bytes.Buffer
	writer := multipart.NewWriter(&body)

	if spaceID != "" {
		if err := writer.WriteField("spaceId", spaceID); err != nil {
			return nil, err
		}
	}
	if err := writer.WriteField("type", string(params.Type)); err != nil {
		return nil, err
	}
	taskParams := params.Params
	if taskParams == nil {
		taskParams = map[string]any{}
	}
	paramsJSON, err := json.Marshal(taskParams)
	if err != nil {
		return nil, err
	}
	if err := writer.WriteField("params", string(paramsJSON)); err != nil {
		return nil, err
	}
	if params.Name != "" {
		if err := writer.WriteField("name", params.Name); err != nil {
			return nil, err
		}
	}
	for _, file := range files {
		part, err := writer.CreateFormFile("files", file.Name)
		if err != nil {
			return nil, err
		}
		if _, err := part.Write(file.Data); err != nil {
			return nil, err
		}
		if params.Upload.OnProgress != nil {
			params.Upload.OnProgress(1)
		}
	}
	if err := writer.Close(); err != nil {
		return nil, err
	}

	headers := http.Header{}
	headers.Set("Content-Type", writer.FormDataContentType())
	var task Task
	if _, err := t.http.postMultipart(ctx, "/tools/tasks", headers, body.Bytes(), &task); err != nil {
		return nil, err
	}
	return &task, nil
}

func (t *TasksClient) List(ctx context.Context, params ListTasksParams) (*TaskListResponse, error) {
	spaceID, err := t.http.ResolveSpaceID(ctx, params.SpaceID)
	if err != nil {
		return nil, err
	}
	query := url.Values{}
	if spaceID != "" {
		query.Set("spaceId", spaceID)
	}
	if params.HasStart {
		query.Set("_startIndex", strconv.Itoa(params.StartIndex))
	} else {
		query.Set("_startIndex", "0")
	}
	if params.HasMax {
		query.Set("_maxResults", strconv.Itoa(params.MaxResults))
	} else {
		query.Set("_maxResults", "50")
	}
	if params.Type != "" {
		query.Set("type", string(params.Type))
	}

	var tasks []Task
	res, err := t.http.getJSON(ctx, "/tools/tasks", query, nil, &tasks)
	if err != nil {
		return nil, err
	}
	total := len(tasks)
	if raw := res.Header.Get("X-Content-Record-Total"); raw != "" {
		if parsed, err := strconv.Atoi(raw); err == nil {
			total = parsed
		}
	}
	return &TaskListResponse{Tasks: tasks, Total: total}, nil
}

func (t *TasksClient) Get(ctx context.Context, taskID string, useEventStream bool) (*Task, error) {
	headers := http.Header{}
	if useEventStream {
		headers.Set("response-event-stream", "yes")
	}
	res, err := t.http.getJSON(ctx, "/tools/tasks/"+urlPathEscape(taskID), t.taskQueryParams(), headers, nil)
	if err != nil {
		return nil, err
	}
	if isEventStream(res.Header) {
		return firstTaskFromSSEBytes(res.Body)
	}
	var task Task
	if err := json.Unmarshal(res.Body, &task); err != nil {
		return nil, err
	}
	return &task, nil
}

func (t *TasksClient) Delete(ctx context.Context, taskID string) error {
	return t.http.delete(ctx, "/tools/tasks/"+urlPathEscape(taskID), t.taskQueryParams())
}

// Start triggers an already-created task to begin executing.
//
// Required for guest-mode tasks: the backend creates them in a pending state
// and waits for an explicit PUT /tools/tasks/:id/start before running.
// Authenticated tasks are started automatically by the backend.
func (t *TasksClient) Start(ctx context.Context, taskID string) (*Task, error) {
	var task Task
	if _, err := t.http.putJSON(ctx, "/tools/tasks/"+urlPathEscape(taskID)+"/start", t.taskQueryParams(), nil, &task); err != nil {
		return nil, err
	}
	return &task, nil
}

func (t *TasksClient) Down(ctx context.Context, taskID string, options TaskDownloadOptions, out any) error {
	query := url.Values{}
	if options.Type != "" {
		query.Set("_type", options.Type)
	}
	var q url.Values
	if len(query) > 0 {
		q = query
	}
	_, err := t.http.getJSON(ctx, "/tools/tasks/"+urlPathEscape(taskID)+"/download", q, nil, out)
	return err
}

func (t *TasksClient) Wait(ctx context.Context, taskID string, options WaitForTaskOptions) (*Task, error) {
	timeout := options.Timeout
	if timeout == 0 {
		timeout = DefaultTimeout
	}
	startedAt := time.Now()

	// Fast path: task may already be terminal before SSE connects (common for
	// quick guest-mode tasks after an explicit start).
	current, err := t.Get(ctx, taskID, false)
	if err != nil {
		return nil, err
	}
	if options.OnProgress != nil {
		options.OnProgress(*current)
	}
	switch current.Status {
	case TaskStatusCompleted:
		return current, nil
	case TaskStatusFailed:
		return nil, fmt.Errorf("task failed: %s", firstNonEmpty(current.Error, "Unknown error"))
	}

	remaining := timeout - time.Since(startedAt)
	if remaining <= 0 {
		return nil, fmt.Errorf("task %s did not complete within %s", taskID, timeout)
	}
	ctx, cancel := context.WithTimeout(ctx, remaining)
	defer cancel()

	if !options.DisableSSE {
		task, err := t.waitWithEventStream(ctx, taskID, options.OnProgress)
		if err == nil {
			return task, nil
		}
	}

	interval := options.PollInterval
	if interval == 0 {
		interval = DefaultPollInterval
	}
	return t.waitWithPolling(ctx, taskID, interval, options.OnProgress)
}

func (t *TasksClient) Subscribe(ctx context.Context, taskID string, handlers SubscribeTaskHandlers) (func(), error) {
	ctx, cancel := context.WithCancel(ctx)
	go func() {
		retryCount := 0
		for {
			err := t.openTaskEventStream(ctx, taskID, handlers)
			if err == nil || ctx.Err() != nil {
				return
			}
			if !isSSETransportError(err) || retryCount >= sseMaxRetries {
				if handlers.OnError != nil {
					handlers.OnError(err)
				}
				return
			}
			retryCount++
			if err := sleepContext(ctx, sseRetryInterval); err != nil {
				return
			}
		}
	}()
	return cancel, nil
}

func (t *TasksClient) prepareTaskFiles(ctx context.Context, params CreateTaskParams) ([]*normalizedUpload, error) {
	if len(params.Files) == 0 {
		return nil, nil
	}
	if t.files == nil {
		return nil, fmt.Errorf("file upload is not available for this task client")
	}
	prepared := make([]*normalizedUpload, 0, len(params.Files))
	for _, file := range params.Files {
		options := mergeUploadOptions(params.Upload, file.Options)
		normalized, err := t.files.Prepare(ctx, file.Input, options)
		if err != nil {
			return nil, err
		}
		prepared = append(prepared, normalized)
	}
	return prepared, nil
}

func (t *TasksClient) resolveFileIDs(ctx context.Context, spaceID string, params CreateTaskParams, prepared []*normalizedUpload) ([]string, error) {
	fileIDs := append([]string(nil), params.FileIDs...)
	for i, file := range prepared {
		if t.files == nil {
			return nil, fmt.Errorf("file upload is not available for this task client")
		}
		options := mergeUploadOptions(params.Upload, params.Files[i].Options)
		options.SpaceID = spaceID
		result, err := t.files.UploadPrepared(ctx, file, options)
		if err != nil {
			return nil, err
		}
		fileIDs = append(fileIDs, result.ID)
	}
	return fileIDs, nil
}

func (t *TasksClient) openTaskEventStream(ctx context.Context, taskID string, handlers SubscribeTaskHandlers) error {
	headers := http.Header{}
	headers.Set("response-event-stream", "yes")
	res, err := t.http.eventStream(ctx, "/tools/tasks/"+urlPathEscape(taskID), t.taskQueryParams(), headers)
	if err != nil {
		return err
	}
	defer res.Stream.Close()

	if !isEventStream(res.Header) {
		if strings.Contains(strings.ToLower(res.Header.Get("Content-Type")), "application/json") {
			var task Task
			data, err := io.ReadAll(res.Stream)
			if err != nil {
				return err
			}
			if err := json.Unmarshal(data, &task); err != nil {
				return err
			}
			if handlers.OnUpdate != nil {
				handlers.OnUpdate(task)
			}
			return nil
		}
		return fmt.Errorf("unexpected Content-Type: %s", res.Header.Get("Content-Type"))
	}

	return parseSSE(res.Stream, func(data string) bool {
		var task Task
		if err := json.Unmarshal([]byte(data), &task); err != nil {
			if handlers.OnError != nil {
				handlers.OnError(err)
			}
			return false
		}
		if handlers.OnUpdate != nil {
			handlers.OnUpdate(task)
		}
		return task.Status == TaskStatusCompleted || task.Status == TaskStatusFailed
	})
}

func (t *TasksClient) waitWithEventStream(ctx context.Context, taskID string, onProgress func(Task)) (*Task, error) {
	type waitResult struct {
		task *Task
		err  error
	}
	done := make(chan waitResult, 1)
	send := func(result waitResult) {
		select {
		case done <- result:
		default:
		}
	}
	cancel, err := t.Subscribe(ctx, taskID, SubscribeTaskHandlers{
		OnUpdate: func(task Task) {
			if onProgress != nil {
				onProgress(task)
			}
			if task.Status == TaskStatusCompleted {
				send(waitResult{task: &task})
			} else if task.Status == TaskStatusFailed {
				send(waitResult{err: fmt.Errorf("task failed: %s", firstNonEmpty(task.Error, "Unknown error"))})
			}
		},
		OnError: func(err error) {
			send(waitResult{err: err})
		},
	})
	if err != nil {
		return nil, err
	}
	defer cancel()

	select {
	case result := <-done:
		return result.task, result.err
	case <-ctx.Done():
		return nil, ctx.Err()
	}
}

func (t *TasksClient) waitWithPolling(ctx context.Context, taskID string, interval time.Duration, onProgress func(Task)) (*Task, error) {
	for {
		task, err := t.Get(ctx, taskID, false)
		if err != nil {
			return nil, err
		}
		if onProgress != nil {
			onProgress(*task)
		}
		switch task.Status {
		case TaskStatusCompleted:
			return task, nil
		case TaskStatusFailed:
			return nil, fmt.Errorf("task failed: %s", firstNonEmpty(task.Error, "Unknown error"))
		}
		if err := sleepContext(ctx, interval); err != nil {
			return nil, err
		}
	}
}

func (t *TasksClient) taskQueryParams() url.Values {
	spaceID := t.http.SpaceID()
	if spaceID == "" {
		return nil
	}
	query := url.Values{}
	query.Set("spaceId", spaceID)
	return query
}

func parseSSE(r io.Reader, onEvent func(data string) bool) error {
	scanner := bufio.NewScanner(r)
	scanner.Buffer(make([]byte, 0, 64*1024), 10*1024*1024)
	var dataLines []string
	flush := func() bool {
		if len(dataLines) == 0 {
			return false
		}
		data := strings.Join(dataLines, "\n")
		dataLines = nil
		return onEvent(data)
	}
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			if flush() {
				return nil
			}
			continue
		}
		if strings.HasPrefix(line, "data:") {
			dataLines = append(dataLines, strings.TrimPrefix(strings.TrimPrefix(line, "data:"), " "))
		}
	}
	if err := scanner.Err(); err != nil {
		return err
	}
	if flush() {
		return nil
	}
	return fmt.Errorf("SSE connection ended before task completion")
}

func firstTaskFromSSEBytes(data []byte) (*Task, error) {
	var found *Task
	err := parseSSE(strings.NewReader(string(data)), func(raw string) bool {
		var task Task
		if json.Unmarshal([]byte(raw), &task) == nil {
			found = &task
			return true
		}
		return false
	})
	if found != nil {
		return found, nil
	}
	return nil, err
}

func isSSETransportError(err error) bool {
	if err == nil {
		return false
	}
	var apiErr *APIError
	if errors.As(err, &apiErr) {
		return false
	}
	return !strings.HasPrefix(err.Error(), "unexpected Content-Type:")
}

func mergeUploadOptions(defaults UploadOptions, override UploadOptions) UploadOptions {
	out := defaults
	if override.SpaceID != "" {
		out.SpaceID = override.SpaceID
	}
	if override.Name != "" {
		out.Name = override.Name
	}
	if override.Hash != "" {
		out.Hash = override.Hash
	}
	if override.ChunkSize != 0 {
		out.ChunkSize = override.ChunkSize
	}
	if override.OnProgress != nil {
		out.OnProgress = override.OnProgress
	}
	return out
}

func urlPathEscape(value string) string {
	return url.PathEscape(value)
}

File: sdks/python/decktools/tasks.py
from __future__ import annotations

import json
import threading
import time
from collections.abc import Callable, Iterator, Mapping, MutableSequence, Sequence
from concurrent.futures import ThreadPoolExecutor
from typing import Any, cast
from urllib.parse import quote

import httpx

from .errors import APIError
from .files import FilesClient, _NormalizedUpload
from .http_client import HTTPClient
from .types import (
    DEFAULT_POLL_INTERVAL,
    DEFAULT_TIMEOUT,
    INLINE_TASK_FILES_MAX_BYTES,
    ErrorCallback,
    Task,
    TaskCallback,
    TaskListResponse,
    TaskResult,
    TaskUploadInput,
)

SSE_RETRY_INTERVAL = 5.0
SSE_MAX_RETRIES = 100
_TERMINAL_STATUSES = frozenset(("completed", "failed"))


class TasksClient:
    def __init__(self, http: HTTPClient, files: FilesClient | None = None) -> None:
        self._http = http
        self._files = files

    def create(
        self,
        task_type: str | None = None,
        *,
        type: str | None = None,
        space_id: str | None = None,
        file_ids: Sequence[str] | None = None,
        files: Sequence[TaskUploadInput] | None = None,
        name: str | None = None,
        params: Mapping[str, Any] | None = None,
        upload: Mapping[str, Any] | None = None,
    ) -> Task:
        resolved_type = type or task_type
        if not resolved_type:
            raise ValueError("task type is required")
        resolved_space_id = self._http.resolve_space_id(space_id)
        resolved_file_ids = list(file_ids or ())
        prepared: list[_NormalizedUpload] = []
        if files:
            if self._files is None:
                raise RuntimeError("file upload is not available for this task client")
            files_client = self._files
            defaults = dict(upload or {})

            def prepare_one(item: TaskUploadInput) -> _NormalizedUpload:
                source, options = self._normalize_task_upload(item, defaults)
                prepare_options = {
                    key: value
                    for key, value in options.items()
                    if key in {"name", "hash", "chunk_size"}
                }
                return files_client.prepare(source, **prepare_options)

            with ThreadPoolExecutor(max_workers=min(5, len(files))) as pool:
                prepared.extend(pool.map(prepare_one, files))

        total_bytes = sum(item.size for item in prepared)
        can_inline = (
            bool(prepared)
            and not resolved_file_ids
            and total_bytes < INLINE_TASK_FILES_MAX_BYTES
        )
        if can_inline:
            return self._create_with_inline_files(
                resolved_space_id,
                resolved_type,
                prepared,
                name=name,
                params=params,
                upload=upload,
            )

        if prepared:
            if self._files is None:
                raise RuntimeError("file upload is not available for this task client")
            files_client = self._files
            defaults = dict(upload or {})

            def upload_one(item: tuple[_NormalizedUpload, TaskUploadInput]) -> str:
                normalized, source = item
                _, options = self._normalize_task_upload(source, defaults)
                on_progress = options.get("on_progress")
                result = files_client.upload_prepared(
                    normalized,
                    space_id=resolved_space_id,
                    on_progress=on_progress,
                )
                return result["id"]

            with ThreadPoolExecutor(max_workers=min(5, len(prepared))) as pool:
                resolved_file_ids.extend(
                    pool.map(upload_one, zip(prepared, files or (), strict=True))
                )

        payload: dict[str, Any] = {
            "spaceId": resolved_space_id,
            "fileIds": resolved_file_ids,
            "type": resolved_type,
            "params": dict(params or {}),
        }
        if name:
            payload["name"] = name
        response = self._http.request("POST", "/tools/tasks", json=payload)
        result = response.json()
        if not isinstance(result, dict):
            raise TypeError("task creation response must be an object")
        return result

    def _create_with_inline_files(
        self,
        space_id: str,
        task_type: str,
        files: Sequence[_NormalizedUpload],
        *,
        name: str | None,
        params: Mapping[str, Any] | None,
        upload: Mapping[str, Any] | None,
    ) -> Task:
        form_fields: dict[str, str] = {
            "spaceId": space_id,
            "type": task_type,
            "params": json.dumps(dict(params or {})),
        }
        if name:
            form_fields["name"] = name

        multipart_files = [("files", (file.name, file.data)) for file in files]
        on_progress = (upload or {}).get("onProgress") or (upload or {}).get("on_progress")
        if callable(on_progress):
            for _ in files:
                on_progress(1.0)

        response = self._http.request(
            "POST",
            "/tools/tasks",
            data=form_fields,
            files=multipart_files,
        )
        result = response.json()
        if not isinstance(result, dict):
            raise TypeError("task creation response must be an object")
        return result

    @staticmethod
    def _normalize_task_upload(
        item: TaskUploadInput,
        defaults: dict[str, Any],
    ) -> tuple[Any, dict[str, Any]]:
        if isinstance(item, Mapping) and "input" in item:
            source = item["input"]
            options = {**defaults, **{key: value for key, value in item.items() if key != "input"}}
        else:
            source = item
            options = dict(defaults)

        aliases = {
            "spaceId": "space_id",
            "chunkSize": "chunk_size",
            "onProgress": "on_progress",
        }
        return source, {aliases.get(key, key): value for key, value in options.items()}

    def list(
        self,
        *,
        space_id: str | None = None,
        type: str | None = None,
        start_index: int = 0,
        max_results: int = 50,
    ) -> TaskListResponse:
        resolved_space_id = self._http.resolve_space_id(space_id)
        query: dict[str, Any] = {
            "spaceId": resolved_space_id,
            "_startIndex": start_index,
            "_maxResults": max_results,
        }
        if type:
            query["type"] = type
        response = self._http.request("GET", "/tools/tasks", params=query)
        tasks = response.json()
        if not isinstance(tasks, list):
            raise TypeError("task list response must be an array")
        raw_total = response.headers.get("x-content-record-total")
        try:
            total = int(raw_total) if raw_total is not None else len(tasks)
        except ValueError:
            total = len(tasks)
        return {"tasks": tasks, "total": total}

    def get(self, task_id: str, *, use_event_stream: bool = False) -> Task:
        headers = {"response-event-stream": "yes"} if use_event_stream else None
        response = self._http.request(
            "GET",
            f"/tools/tasks/{quote(task_id, safe='')}",
            params=self._task_query_params(),
            headers=headers,
        )
        content_type = response.headers.get("content-type", "").lower()
        if "event-stream" in content_type:
            task = next(self._tasks_from_sse_lines(response.text.splitlines()), None)
            if task is None:
                raise ValueError("event stream did not contain a valid task")
            return task
        result = response.json()
        if not isinstance(result, dict):
            raise TypeError("task detail response must be an object")
        return result

    def delete(self, task_id: str) -> None:
        self._http.request(
            "DELETE",
            f"/tools/tasks/{quote(task_id, safe='')}",
            params=self._task_query_params(),
        )

    def down(
        self,
        task_id: str,
        *,
        type: str | None = None,
    ) -> TaskResult:
        query = {"_type": type} if type else None
        response = self._http.request(
            "GET",
            f"/tools/tasks/{quote(task_id, safe='')}/download",
            params=query,
        )
        return response.json()

    def wait(
        self,
        task_id: str,
        *,
        timeout: float = DEFAULT_TIMEOUT,
        use_event_stream: bool = True,
        on_progress: TaskCallback | None = None,
        poll_interval: float = DEFAULT_POLL_INTERVAL,
    ) -> Task:
        started = time.monotonic()
        if use_event_stream:
            done = threading.Event()
            state: dict[str, Any] = {}

            def on_update(task: Task) -> None:
                if on_progress:
                    on_progress(task)
                status = task.get("status")
                if status == "completed":
                    state["task"] = task
                    done.set()
                elif status == "failed":
                    state["error"] = RuntimeError(
                        f"Task failed: {task.get('error') or 'Unknown error'}"
                    )
                    done.set()

            def on_error(error: Exception) -> None:
                state["stream_error"] = error
                done.set()

            cancel = self.subscribe(task_id, on_update=on_update, on_error=on_error)
            if done.wait(timeout):
                cancel()
                if "task" in state:
                    return cast(Task, state["task"])
                if "error" in state:
                    raise cast(Exception, state["error"])
                remaining = timeout - (time.monotonic() - started)
                if remaining <= 0:
                    raise TimeoutError(f"Task {task_id} did not complete within {timeout}s")
                return self._wait_with_polling(
                    task_id,
                    timeout=remaining,
                    poll_interval=poll_interval,
                    on_progress=on_progress,
                )
            cancel()
            raise TimeoutError(f"Task {task_id} did not complete within {timeout}s")

        return self._wait_with_polling(
            task_id,
            timeout=timeout,
            poll_interval=poll_interval,
            on_progress=on_progress,
        )

    def subscribe(
        self,
        task_id: str,
        *,
        on_update: TaskCallback,
        on_error: ErrorCallback | None = None,
    ) -> Callable[[], None]:
        canceled = threading.Event()
        active_lock = threading.Lock()
        active_response: MutableSequence[httpx.Response | None] = [None]

        def close_active() -> None:
            with active_lock:
                if active_response[0] is not None:
                    active_response[0].close()
                    active_response[0] = None

        def run() -> None:
            retries = 0
            while not canceled.is_set():
                try:
                    self._open_event_stream(
                        task_id,
                        on_update=on_update,
                        on_parse_error=on_error,
                        canceled=canceled,
                        active_response=active_response,
                        active_lock=active_lock,
                    )
                    return
                except Exception as error:
                    if canceled.is_set():
                        return
                    if not self._is_sse_transport_error(error) or retries >= SSE_MAX_RETRIES:
                        if on_error:
                            on_error(error)
                        return
                    retries += 1
                    if canceled.wait(SSE_RETRY_INTERVAL):
                        return

        thread = threading.Thread(target=run, name=f"decktools-sse-{task_id}", daemon=True)
        thread.start()

        def cancel() -> None:
            canceled.set()
            close_active()

        return cancel

    def _open_event_stream(
        self,
        task_id: str,
        *,
        on_update: TaskCallback,
        on_parse_error: ErrorCallback | None,
        canceled: threading.Event,
        active_response: MutableSequence[httpx.Response | None],
        active_lock: threading.Lock,
    ) -> None:
        with self._http.stream(
            f"/tools/tasks/{quote(task_id, safe='')}",
            params=self._task_query_params(),
            headers={"response-event-stream": "yes"},
        ) as response:
            with active_lock:
                active_response[0] = response
            try:
                content_type = response.headers.get("content-type", "").lower()
                if "application/json" in content_type:
                    response.read()
                    payload = response.json()
                    if not isinstance(payload, dict):
                        raise TypeError("task stream JSON response must be an object")
                    on_update(payload)
                    return
                if "event-stream" not in content_type:
                    raise ValueError(f"Unexpected Content-Type: {content_type}")

                for task in self._tasks_from_sse_lines(response.iter_lines(), on_parse_error):
                    if canceled.is_set():
                        return
                    on_update(task)
                    if task.get("status") in _TERMINAL_STATUSES:
                        return
                if not canceled.is_set():
                    raise ConnectionError("SSE connection ended before task completion")
            finally:
                with active_lock:
                    active_response[0] = None

    @staticmethod
    def _tasks_from_sse_lines(
        lines: Iterator[str] | Sequence[str],
        on_parse_error: ErrorCallback | None = None,
    ) -> Iterator[Task]:
        data_lines: list[str] = []

        def parse() -> Task | None:
            if not data_lines:
                return None
            raw = "\n".join(data_lines)
            data_lines.clear()
            try:
                task = json.loads(raw)
            except (TypeError, ValueError) as error:
                if on_parse_error:
                    on_parse_error(error)
                return None
            return task if isinstance(task, dict) else None

        for line in lines:
            if line == "":
                task = parse()
                if task is not None:
                    yield task
                continue
            if line.startswith("data:"):
                data_lines.append(line[5:].lstrip(" "))
        task = parse()
        if task is not None:
            yield task

    @staticmethod
    def _is_sse_transport_error(error: Exception) -> bool:
        if isinstance(error, APIError):
            return False
        return not (
            isinstance(error, ValueError) and str(error).startswith("Unexpected Content-Type:")
        )

    def _wait_with_polling(
        self,
        task_id: str,
        *,
        timeout: float,
        poll_interval: float,
        on_progress: TaskCallback | None,
    ) -> Task:
        started = time.monotonic()
        while True:
            if time.monotonic() - started > timeout:
                raise TimeoutError(f"Task {task_id} did not complete within {timeout}s")
            task = self.get(task_id)
            if on_progress:
                on_progress(task)
            if task.get("status") == "completed":
                return task
            if task.get("status") == "failed":
                raise RuntimeError(f"Task failed: {task.get('error') or 'Unknown error'}")
            time.sleep(poll_interval)

    def _task_query_params(self) -> dict[str, str] | None:
        return {"spaceId": self._http.space_id} if self._http.space_id else None


# TypeScript naming.
TasksApi = TasksClient
